File: ruby/4a.rb
#!/usr/bin/ruby

rooms = File.open('4.in').readlines.map do |x|

  # Extract the room
  r = x.rpartition('-').first

  # Group by characters (frequency count), sort by freqency and alphabet
  c = r.gsub(/-/, '').chars.group_by(&:chr).sort_by{ |x,y| [0 - y.length, x] }

  # Return an array containing computed checksum, sector and checksum
  [ c.map(&:first)[0..4].join, x.scan(/\d+/).first.to_i, x.scan(/\w+/).last]

end.select{ |x| x.first == x.last }

puts rooms.map{ |x| x[1] }.reduce(:+)

File: python/4a.py
#!/usr/bin/python

import collections

def main():
    result = 0
    for line in open('4.in'):
        # Extract the room and extra data
        room, extra = line.rsplit('-', 1)
        # Count the occurrences of letters ignoring hyphens
        counts = collections.Counter(room.replace('-', ''))
        # Sort based on highest requency then lexically
        ordered = sorted(counts.items(), key=lambda x: (-x[1], x[0]))
        # The checksum is the first 5 characters from our ordered list
        computed = ''.join(x for x, y in ordered[:5])
        # Extract the sector and checksum
        sector, checksum = extra[:-2].split('[')
        # Ignore faulty checksums
        if checksum != computed:
            continue
        # Increment the sector counter
        result += int(sector)
    print result

if __name__ == '__main__':
    main()

# vi: ts=4 et:
